### PR TITLE
Crafting fields & properties

### LINKED ISSUE #856: Sync: Crafting Fields
## Description

1. Sync the values in the **Synced** list below.
2. Create an end to end (E2E) tests for all fields (1 for server changing the value and 1 for client changing the value if applicable)
3. Create a command to test changing each value in game [see similar command](https://github.com/Bannerlord-Coop-Team/BannerlordCoop/blob/development/source/GameInterface/Services/Towns/Commands/TownDebugCommand.cs#L455)

Also see additional information at the bottom for more information and templates

**Key**
Server side only (only allow the running of the function on the server side)
Client side only (only allow the running of the function on the client side)
Client side sync (request the server to change the value, server then changes the value and replies to the client allowing the change)
Server side sync (only server allows running of the function and send to all clients the value changed and clients update the value/call original function)

**Synced**

| Method/Field Name | Sync Type | Additional Information |
|:-----------------|:---------:|:--------------|
| _currentHistoryIndex | Server side sync | |
| _craftedItemObject | Server side sync | |

**Deferred**

None

**Externally Deferred**

None

**Non-Synced**

None

## Intended Design

For the patch itself please refer to https://github.com/Bannerlord-Coop-Team/BannerlordCoop/wiki/Patching-Basics#field-patching

**Field Sync Dataflow**
FieldSyncDataflow
**Integration Testing**
image
## Location

Create a branch based from [development](https://github.com/Bannerlord-Coop-Team/BannerlordCoop)

## Related Issues

Blocked by:
- #855

## Requirements

N/A

## Additional information

Using the [GameInterface Service Tempale](https://github.com/Bannerlord-Coop-Team/BannerlordCoop/tree/development/source/GameInterface/Services/Template) create commands to test the synced values.

## Definition of Done
- [ ] Class level comments exist for all new classes.
- [ ] XUnit E2E tests exist for every property on the server (and client if applicable)
- [ ] Each field has a command getter and setter that used the correct intercept

### LINKED ISSUE #857: Sync: Crafting Properties
## Description

1. Sync the values in the **Synced** list below.
2. Create an end to end (E2E) tests for all properties (1 for server changing the value and 1 for client changing the value if applicable)
3. Create a command to test changing each property in game [see similar command](https://github.com/Bannerlord-Coop-Team/BannerlordCoop/blob/development/source/GameInterface/Services/Towns/Commands/TownDebugCommand.cs#L455)

Also see additional information at the bottom for more information and templates

**Key**
Server side only (only allow the running of the function on the server side)
Client side only (only allow the running of the function on the client side)
Client side sync (request the server to change the value, server then changes the value and replies to the client allowing the change)
Server side sync (only server allows running of the function and send to all clients the value changed and clients update the value/call original function)

**Synced**

| Method/Field Name | Sync Type | Additional Information |
|:-----------------|:---------:|:--------------|
| CurrentWeaponDesign | Server side sync | |
| CurrentItemModifierGroup | Server side sync | |
| CraftedWeaponName | Server side sync | |

**Deferred**

None

**Externally Deferred**

None

**Non-Synced**

None

## Intended Design

You are expected to only sync the property setters.
Try using the new [autosync tool](https://github.com/Bannerlord-Coop-Team/BannerlordCoop/wiki/Property-AutoSync)

## Location

Create a branch based from [development](https://github.com/Bannerlord-Coop-Team/BannerlordCoop)

## Related Issues

Blocked by:
- #855

## Requirements

N/A

## Additional information

Using the [GameInterface Service Tempale](https://github.com/Bannerlord-Coop-Team/BannerlordCoop/tree/development/source/GameInterface/Services/Template) create commands to test the synced values.

## Definition of Done
- [ ] Class level comments exist for all new classes.
- [ ] XUnit E2E tests exist for every property on the server (and client if applicable)
- [ ] Each property has a command getter and setter

### LINKED ISSUE #993: ItemObject Lifetime Sync
## Description

1. Sync the values in the **Synced** list below.
2. Create an integration test for all values (1 for server changing the value and 1 for client changing the value if applicable)
3. Create a command to test changing the value in game [see similar command](https://github.com/Bannerlord-Coop-Team/BannerlordCoop/blob/development/source/GameInterface/Services/Towns/Commands/TownDebugCommand.cs#L455)

Also see additional information at the bottom for more information and templates

**Synced**

| Method/Field Name | Sync Type |
|:-----------------:|:---------:|
| TODO              | TODO |

**Deferred**

None

**Externally Deferred**

None

**Non-Synced**
None

## Intended Design

**Message networking**
image
**Integration Testing**
image
## Location

Create a branch based from [development](https://github.com/Bannerlord-Coop-Team/BannerlordCoop)

## Related Issues

N/A

## Requirements

N/A

## Additional information

Using the [GameInterface Service Tempale](https://github.com/Bannerlord-Coop-Team/BannerlordCoop/tree/development/source/GameInterface/Services/Template) create commands to test the synced values.

For network communication use [Server Service Template](https://github.com/Bannerlord-Coop-Team/BannerlordCoop/tree/development/source/Coop.Core/Server/Services/Template) and [Client Service Template](https://github.com/Bannerlord-Coop-Team/BannerlordCoop/tree/development/source/Coop.Core/Client/Services/Template)

It is recommended to start with an integration test, you can create one using the [test template](https://github.com/Bannerlord-Coop-Team/BannerlordCoop/tree/development/source/Coop.IntegrationTests/Template).

## Definition of Done
- [ ] Class level comments exist for all new classes.
- [ ] XUnit tests exist for every method that does not require the game to be ran.
- [ ] Commands exist for all new sync items.

### PR DESCRIPTION
**Pull request by EgardA**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] All new classes have class-level documentation comments, if there are any at all
- [x] Tests for the changes have been added (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation update
- [ ] Other... Please describe:

## What is the new behavior?

Resolves #857 
Resolves #856

### CODE REVIEW

File: source/E2E.Tests/Services/CraftingService/CraftingSyncTests.cs [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+                serverCrafting.CraftedWeaponName = new TextObject("craftedWeapon");
+                WeaponDesign weaponDesign = ObjectHelper.SkipConstructor<WeaponDesign>();
+                serverCrafting.CurrentWeaponDesign = weaponDesign;
+            });
+
+            foreach (var client in TestEnvironment.Clients)
+            {
+                Assert.True(client.ObjectManager.TryGetObject(craftingId, out Crafting clientCrafting));
+                Assert.Equal(serverCrafting._currentHistoryIndex, clientCrafting._currentHistoryIndex);
+                Assert.Equal(serverCrafting._craftedItemObject, clientCrafting._craftedItemObject);
```
Comment (garrettluskey) [inline source/E2E.Tests/Services/CraftingService/CraftingSyncTests.cs]: item object should have it's own handler and be treated as a network object

**Review by garrettluskey — CHANGES_REQUESTED** [review]

File: source/E2E.Tests/Services/CraftingService/CraftingSyncTests.cs (hunk 1 above)
Comment (EgardA) [reply source/E2E.Tests/Services/CraftingService/CraftingSyncTests.cs]: Yeah, but that's a different issue. Or am I misunderstanding you?

Comment (garrettluskey) [reply source/E2E.Tests/Services/CraftingService/CraftingSyncTests.cs]: This is asserting the values are equal rather than a network string id

Comment (EgardA) [reply source/E2E.Tests/Services/CraftingService/CraftingSyncTests.cs]: Fair

File: source/GameInterface/Services/WeaponDesigns/WeaponDesignRegistry.cs [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    internal class WeaponDesignRegistry : RegistryBase<WeaponDesign>
+    {
+        private const string ItemObjectIdPrefix = "CoopWeaponDesign";
+        private static int InstanceCounter = 0;
+
+        public WeaponDesignRegistry(IRegistryCollection collection) : base(collection) { }
+
+        public override void RegisterAll()
+        {
+            //TODO
```
Comment (garrettluskey) [inline source/GameInterface/Services/WeaponDesigns/WeaponDesignRegistry.cs]: Do ＞：）

File: source/E2E.Tests/Services/CraftingService/CraftingSyncTests.cs [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+                Assert.True(server.ObjectManager.TryGetId(crafting, out craftingId));
+            });
+
+            // Assert
+            Assert.True(server.ObjectManager.TryGetObject(craftingId, out Crafting serverCrafting));
+
+            server.Call(() =>
+            {
+                currentHistoryIntercept.Invoke(null, new object[] { serverCrafting, 69 });
+                craftingItemObjectIntercept.Invoke(null, new object[] { serverCrafting, new ItemObject("item") });
```
Comment (garrettluskey) [inline source/E2E.Tests/Services/CraftingService/CraftingSyncTests.cs:60]: This isn't properly tested, need to confirm that the ItemObject is the same across the network. Requires syncing of ItemObject

File: source/GameInterface/Services/ItemObjects/Handlers/ItemObjectLifetimeHandler.cs [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        private void Handle_BuildingCreated(MessagePayload<ItemObjectCreated> obj)
+        {
+            var payload = obj.What;
+
+            if (objectManager.AddNewObject(payload.ItemObject, out string itemObjectId) == false) return;
+
+            var message = new NetworkCreateItemObject(itemObjectId);
+            network.SendAll(message);
+        }
+
```
Comment (garrettluskey) [inline source/GameInterface/Services/ItemObjects/Handlers/ItemObjectLifetimeHandler.cs]: This is in fact, not a building

File: source/GameInterface/Services/ItemObjects/ItemObjectRegistry.cs [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    internal class ItemObjectRegistry : RegistryBase<ItemObject>
+    {
+        private const string ItemObjectIdPrefix = "CoopItemObject";
+        private static int InstanceCounter = 0;
+
+        public ItemObjectRegistry(IRegistryCollection collection) : base(collection) { }
+
+        public override void RegisterAll()
+        {
+            //Are these registered somewhere?
```
Comment (garrettluskey) [inline source/GameInterface/Services/ItemObjects/ItemObjectRegistry.cs]:
```suggestion
```

**Review by garrettluskey — CHANGES_REQUESTED** [review]

File: source/E2E.Tests/Services/CraftingService/CraftingSyncTests.cs (hunk 3 above)
Comment (EgardA) [reply source/E2E.Tests/Services/CraftingService/CraftingSyncTests.cs:60]: Shouldn't this test only test that they have the same network id? The test for ItemObject values across the network would be checked in the ItemObject Sync Test?

Comment (garrettluskey) [reply source/E2E.Tests/Services/CraftingService/CraftingSyncTests.cs:60]: Agreed, but where is that test?

**Pull request merged**